Autopilot/PathManager/Inc/waypointManager.hpp

```diff
@@ -0,0 +1,175 @@
+    long double longitude;            // Longitude of waypoint
+    int altitude;                     // Altitude of waypoint
+    float turnRadius;                 // If turn is commanded, this is the radius of the turn
+    char type;
+    char waypointId;                  // Id of the waypoint
+    char arrayIndex;                  // Position of the waypoint within the waypointBuffer (array)
+} _WaypointData;
+
+/*
+* Structure contains the data that will be returned to the Path Manager state manager. 
+* This data will be used by the PID and coordinated turn engine to determine the commands to be sent to the Attitude Manager.
+*/
+typedef struct {
+    uint16_t desiredHeading;            // Desired heading to stay on path
+    int desiredAltitude;                // Desired altitude at next waypoint
+    long double distanceToNextWaypoint; // Distance to the next waypoint (helps with airspeed PID)
+    _WaypointStatus errorCode;          // Contains error codes
+    bool isDataNew;                     // Notifies PID modules if the data in this structure is new
+    uint32_t timeOfData;                // The time that the data in this structure was collected
+} _WaypointManager_Data;
+
+class WaypointManager {
+    public: 
+        /*
+        * Constructor for this class
+        * 
+        * @param _WaypointData * initialWaypoints -> These waypoints will be used to initialize the waypointBuffer array
+        * @param int numWaypoints -> Number of waypoints being initialized in the waypointBuffer array
+        * @param _WaypointData* currentLocation -> Home base. 
+        */
+        WaypointManager(_WaypointData * initialWaypoints, int numWaypoints, _WaypointData* currentLocation);
+        
+        /*
+        * Updates the _WaypointManager_Data structure with new values
+        * 
+        * @param _Gps_Data currentPosition -> contains the current coordinates, altitude, and heading
+        */
+        void get_next_directions(_Gps_Data currentPosition); 
+
+        /*
+        * Adds, inserts, updates, or deletes a single waypoint in the waypointBuffer array
+        * 
+        * @param _WaypointData* waypoint -> In the instance that we need to update, insert, or append a new waypoint, this will be used 
+        * @param int waypointId -> the ID of the waypoint that will be updated or deleted
+        * @param _WaypointBufferUpdateType updateType -> the type of modification to the waypointBuffer array (look above)
+        * @param int previousId -> stores the ID of the waypoint that will come before the inserted waypoint. Set to 0 by default, so does not need to be passed
+        * @param int nextId -> stores the ID of the waypoint that will come after the inserted waypoint. Set to 0 by default, so does not need to be passed
+        */
+        void update_path_nodes(_WaypointData* waypoint, int waypointId, _WaypointBufferUpdateType updateType, int previousId = 0, int nextId = 0);
```

Comment:
confused about the id system. If I just want to add a new way point to the list, what would I do about waypointId ? Does your module expect the caller to keep a list of the way point history ?

Reply:
That's true. When appending a waypoint, the id does not need to be passed in separately. I will change that.

Here's how the ID system works:

After a waypoint is created, it is stored in the waypointBuffer array. However, as waypoints are inserted, deleted, etc. the index of the waypoints will change during the duration of the flight. 

For example, say when you initialized the waypointBuffer array, you put w1 at waypointBuffer[2] and w2 at waypointBuffer[3]. However because of deletions and insertions, w1 and w2 are now found at waypointBuffer[5] and waypointBuffer[6]. Now, say you want to insert a new waypoint, w3, between w1 and w2. You can't simply list the array index value since the state machine has no idea what it is. 

Instead, if we give each waypoint an ID (first waypoint has an ID of 1, second 2, and the nth n), you can pass in the ID of the waypoint you care about and the waypoint manager will have a method to find the index of the waypoint with that ID. The state machine will be aware of the waypoint IDs (because the state machine creates the waypoint objects).

Hope that helps...